File: 0029_Divide_Two_Integers/29.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution
{
public:
    long long int divide(long long int dividend, long long int divisor)
    {
        long long int ans = (long long int)(dividend / divisor);
        // cout << ans;
        if (ans > pow(2, 31) - 1 || ans < pow(-2, 31))
        {

            return pow(2, 31) - 1;
        }
        if (dividend % divisor == 0)
        {

            return ans;
        }
        else
        {
            return ans;
        }
    }
};

int main()
{
    Solution a;
    cout << a.divide(10, 3) << "\n"
         << a.divide(7, -3) << "\n"
         << a.divide(0, 1) << "\n"
         << a.divide(1, 1) << "\n";
    return 0;
}

File: 0029_Divide_Two_Integers/29.py
class Solution:
    def divide(self, dividend: int, divisor: int) -> int:
        ans = dividend // divisor
        if ans < -2 ** 31 or ans > 2 ** 31 - 1:
            return 2 ** 31 - 1
        if dividend % divisor == 0:
            return ans
        else:
            return ans if ans >= 0 else ans + 1

a = Solution()
print(a.divide(dividend = 10, divisor = 3))
print(a.divide(dividend = 7, divisor = -3))
print(a.divide(dividend = 0, divisor = 1))
print(a.divide(dividend = 1, divisor = 1))